File: src/main/java/no/novari/flyt/kafka/instanceflow/producing/InstanceFlowTemplateFactory.java
package no.novari.flyt.kafka.instanceflow.producing;

import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeadersMapper;
import no.novari.kafka.producing.ParameterizedTemplateFactory;
import org.springframework.stereotype.Service;

@Service
public class InstanceFlowTemplateFactory {

    private final ParameterizedTemplateFactory parameterizedTemplateFactory;
    private final InstanceFlowHeadersMapper instanceFlowHeadersMapper;

    public InstanceFlowTemplateFactory(
            ParameterizedTemplateFactory parameterizedTemplateFactory,
            InstanceFlowHeadersMapper instanceFlowHeadersMapper
    ) {
        this.parameterizedTemplateFactory = parameterizedTemplateFactory;
        this.instanceFlowHeadersMapper = instanceFlowHeadersMapper;
    }

    public <VALUE> InstanceFlowTemplate<VALUE> createTemplate(Class<VALUE> valueClass) {
        return new InstanceFlowTemplate<>(
                parameterizedTemplateFactory.createTemplate(valueClass),
                instanceFlowHeadersMapper
        );
    }
}

File: src/main/kotlin/no/novari/flyt/kafka/instanceflow/producing/InstanceFlowTemplateFactory.kt
package no.novari.flyt.kafka.instanceflow.producing

import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeadersMapper
import no.novari.kafka.producing.ParameterizedTemplateFactory
import org.springframework.stereotype.Service

@Service
class InstanceFlowTemplateFactory(
    private val parameterizedTemplateFactory: ParameterizedTemplateFactory,
    private val instanceFlowHeadersMapper: InstanceFlowHeadersMapper,
) {
    fun <VALUE> createTemplate(valueClass: Class<VALUE>): InstanceFlowTemplate<VALUE> =
        InstanceFlowTemplate(
            parameterizedTemplate = parameterizedTemplateFactory.createTemplate(valueClass),
            instanceFlowHeadersMapper = instanceFlowHeadersMapper,
        )
}
